gazebo_ros/src/executor.cpp

```diff
@@ -109,31 +112,38 @@ Executor::run()
 {
   while (rclcpp::ok(this->context_) && spinning.load()) {
     rclcpp::AnyExecutable any_exec;
-    std::lock_guard<std::mutex> wait_lock(wait_mutex_);
-    if (!rclcpp::ok(this->context_) || !spinning.load()) {
-      return;
-    }
+    add_remove_node_high_priority_mutex_.lock();
+    std::lock_guard<std::mutex> add_remove_node_lp_lock(add_remove_node_low_priority_mutex_);
+    std::lock_guard<std::mutex> add_remove_node_lock(add_remove_node_mutex_);
+    add_remove_node_high_priority_mutex_.unlock();
+    {
+      std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+      if (!rclcpp::ok(this->context_) || !spinning.load()) {
+        return;
+      }
 
-    if (!get_next_executable(any_exec, next_exec_timeout_)) {
-      continue;
-    }
+      if (!get_next_executable(any_exec, std::chrono::nanoseconds(0))) {
```

Comment:
timeout must be `-1` not `0` to get blocking behavior, which IIUC is the desired behavior here

Reply:
Right, I meant `-1`. However, by blocking here we hit the same deadlock I was getting before. For example, in the test the following line will block forever because one or more spin threads are retaining a lock on mutex for adding/removing nodes:

https://github.com/ros-simulation/gazebo_ros_pkgs/blob/f9c0b54c4d8b52a3a1cdd23f242daf53b2e88d1b/gazebo_ros/test/test_executor.cpp#L38

This is why I originally had a timeout; I can't see a better way.

Reply:
I see, the thing is that we shouldn't hold that lock while calling `rcl_wait()`, and IIUC we actually don't need to.
We could fix this more cleanly in `rclcpp`, though I think it's acceptable to use a timeout for the moment.

are we planning to fix this in `rclcpp` or is there a reason why we are making the fix here?

Reply:
> are we planning to fix this in rclcpp or is there a reason why we are making the fix here?

I opted to fix it here instead because I'm not sure the how to fix the issue in rclcpp.

Reply:
See https://github.com/ros2/rclcpp/pull/1505.

Could you check if that solves the issue here?